### PR TITLE
Edit timesheet

### PR DESCRIPTION
**Pull request by ajaysudhakaranarse**
- Added UI to edit timesheet.
- Added edit and update action to update timesheet.
-Wrote a test cases.

### CODE REVIEW

File: Gemfile [hunk 1]
```diff
@@ -125,4 +125,5 @@ gem "byebug"
 gem "rest-client"
 gem 'time_difference'
 gem 'slack-ruby-bot'
-
+gem 'momentjs-rails', '>= 2.9.0'
```
Comment (Shwetakale) [inline Gemfile]: Why we need this?

File: Gemfile [hunk 2]
```diff
@@ -125,4 +125,5 @@ gem "byebug"
...
+gem 'bootstrap3-datetimepicker-rails', '~> 4.17.47'
```
Comment (Shwetakale) [inline Gemfile]: We are using bootstrap 2 in project and already have bootstrap 2's date picker then why we need this? And will it not conflict to use bootstrap3's datetimepicker with bootstarp2

File: app/assets/javascripts/application.js [hunk 3]
```diff
@@ -42,6 +43,12 @@ $('.datepicker').datepicker({
 })
 })
 
+$(document).ready(function(){
+$('.datetimepicker1').datetimepicker({
```
Comment (Shwetakale) [inline app/assets/javascripts/application.js]: Please use proper naming convention

File: app/controllers/time_sheets_controller.rb [hunk 4]
```diff
@@ -19,13 +19,40 @@ def index
     @timesheet_report = TimeSheet.generete_employee_timesheet_report(timesheets, @from_date.to_date, @to_date.to_date) if timesheets.present?
   end
 
-  def show
```
Comment (Shwetakale) [inline app/controllers/time_sheets_controller.rb]: You have removed show method. Are we not using it?

File: app/controllers/time_sheets_controller.rb [hunk 5]
```diff
@@ -19,13 +19,40 @@ def index
...
+  def users_timesheet
+    redirect_to request.referrer if current_user.role.in?(['Employee', 'Intern']) && current_user.id.to_s != params[:user_id]
+    
     @from_date = params[:from_date]
-    @to_date = params[:to_date]
-    @user = User.find(params[:id])
+    @to_date   = params[:to_date]
+    @user      = User.find(params[:user_id])
```
Comment (Shwetakale) [inline app/controllers/time_sheets_controller.rb]: Please remove extra spaces

File: app/controllers/time_sheets_controller.rb [hunk 6]
```diff
@@ -19,13 +19,40 @@ def index
...
   end
 
-  def show
+  def users_timesheet
+    redirect_to request.referrer if current_user.role.in?(['Employee', 'Intern']) && current_user.id.to_s != params[:user_id]
+    
     @from_date = params[:from_date]
-    @to_date = params[:to_date]
-    @user = User.find(params[:id])
+    @to_date   = params[:to_date]
```
Comment (Shwetakale) [inline app/controllers/time_sheets_controller.rb]: Please remove extra spaces

File: app/controllers/time_sheets_controller.rb [hunk 7]
```diff
@@ -19,13 +19,40 @@ def index
...
+    @user = User.find_by(id: params[:id])
+    @time_sheets = @user.time_sheets.where(date: params[:time_sheet_date].to_date)
+  end
+
+  def update
+    @from_date = Date.today.beginning_of_month.to_s
+    @to_date = Date.today.to_s
+    @user = User.find_by(id: params['user']['id'])
+    @time_sheet_date = params[:time_sheet_date]
+    return_value = TimeSheet.check_validation_while_updating_time_sheet(update_timesheet_params)
```
Comment (Shwetakale) [inline app/controllers/time_sheets_controller.rb]: cant this be model level validations and will be be handled in valid?

File: app/controllers/time_sheets_controller.rb [hunk 8]
```diff
@@ -19,13 +19,40 @@ def index
...
+    if return_value == true
+      if @user.update_attributes(update_timesheet_params)
+        flash.notice = 'Timesheet Updated Succesfully'
```
Comment (Shwetakale) [inline app/controllers/time_sheets_controller.rb]: At one place you are using ```flash.notice``` and at other place you are using ```flash[:notice]``` please be consistent

File: app/controllers/time_sheets_controller.rb [hunk 9]
```diff
@@ -63,4 +90,8 @@ def user_exists?
   def load_user
     @user = User.where('public_profile.slack_handle' => params['user_id']).first unless params['user_id'].nil?
   end
+  
+  def update_timesheet_params
+    params.require(:user).permit!
```
Comment (Shwetakale) [inline app/controllers/time_sheets_controller.rb]: Cant we permit only required params? Why are we using ```permit!``` 

Are these parameters of user or timehseet?

File: app/models/ability.rb [hunk 10]
```diff
@@ -50,6 +51,7 @@ def employee_abilities(user_id)
     can [:new, :create], LeaveApplication, user_id: user_id
     can [:edit, :update], LeaveApplication, leave_status: 'Pending', user_id: user_id
     can :read, Vendor
+    can [:index], TimeSheet
```
Comment (Shwetakale) [inline app/models/ability.rb]: Can't employee update timesheet?

**Comment by Shwetakale** [comment]
Can we also handle delete as well?

File: app/controllers/time_sheets_controller.rb (hunk 7 above)
Comment (ajaysudhakaranarse) [reply app/controllers/time_sheets_controller.rb]: Yes. But we have already validation method that we was wrote to validate timesheet string when user request from slack. So I am using that methods.

File: app/controllers/time_sheets_controller.rb (hunk 9 above)
Comment (ajaysudhakaranarse) [reply app/controllers/time_sheets_controller.rb]: Ok I go through it and do the changes

File: Gemfile (hunk 2 above)
Comment (ajaysudhakaranarse) [reply Gemfile]: We are using datetimepicker on edit form to show date and time both. we need to discuss on it.

**Pull request merged**